app/controllers/application_controller.rb

```diff
@@ -5,6 +5,142 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "password_security"
   end
 
-end
\ No newline at end of file
+  get '/' do
+    if is_logged_in
+      redirect '/tweets'
+    else
+      erb :homepage
+    end
+  end
+
+  get '/signup' do
+    if is_logged_in
+      redirect '/tweets'
+    else
+      erb :"tweets/users/create_user"
+    end
+  end
+
+  post '/signup' do
+      if params[:username] != "" && params[:email] != "" && params[:password] != ""
+        @user = User.create(username: params[:username], email: params[:email], password: params[:password])
+        session[:user_id] = @user.id
+        redirect '/tweets'
+      else
+        redirect '/signup'
+      end
+  end
+
+  get '/login' do
+    if is_logged_in
+      redirect "/tweets"
+    else
+      erb :"tweets/users/login"
+    end
+  end
+
+  post '/login' do
+    @user = User.find_by(username: params[:username])
+    if @user && @user.authenticate(params[:password])
+       session[:user_id] = @user.id
+       redirect "/tweets"
+     else
+       erb :"tweets/users/login", locals: {message: "Invalid username or password! Please try again."}
+     end
+  end
+
+  get "/logout" do
+    if is_logged_in
+      session.clear
+      redirect '/login'
+    else
+      redirect "/"
+    end
+  end
+
+  get '/tweets' do
+    if is_logged_in
+      @user = current_user
+      erb :"tweets/tweets"
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/tweets/new' do
+    if is_logged_in
+      erb :'/tweets/create_tweet'
+    else
+      redirect '/login'
+    end
+  end
+
+  post '/tweets' do
+    if params["content"] != ""
+      @tweet = Tweet.create(content: params["content"], user_id: session[:user_id])
+      redirect "/tweets/#{@tweet.id}"
+    else
+      redirect "/tweets/new"
+    end
+  end
+
+  get '/tweets/:id' do
+    @id = params[:id]
+    @tweet = Tweet.find(@id)
+    if !is_logged_in
+      redirect "/login"
+    else
+      erb :'/tweets/show_tweet'
+    end
+  end
+
+  get '/tweets/:id/edit' do
+    if is_logged_in
+      @tweet = Tweet.find_by(id: params[:id])
+      erb :'/tweets/edit_tweet'
+    else
+      redirect '/login', locals: {message: "Please login to continue."}
+    end
+  end
+
+  patch '/tweets/:id' do
+    @tweet = Tweet.find(params[:id])
+    if params["content"].size != 0
+      @tweet.content = params["content"]
+      @tweet.save
+      redirect "/tweets/#{@tweet.id}"
+    else
+      redirect "/tweets/#{@tweet.id}/edit"
+    end
+  end
+
+  delete '/tweets/:id/delete' do
+    @user = current_user
+    @tweet = Tweet.find(params[:id])
+    if !current_user.tweets.ids.include?(@tweet.id)
```

Comment:
A slightly more efficient approach here would be:

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.